tiledb/sm/storage_manager/storage_manager.cc

```diff
@@ -1702,17 +1740,97 @@ bool StorageManager::is_vacuum_file(const URI& uri) const {
   return false;
 }
 
+Status StorageManager::get_array_schema_uris(
+    const URI& array_uri,
+    std::vector<URI>* schema_uris,
+    std::vector<URI>* array_ls_uris) const {
+  auto timer_se = stats_->start_timer("read_get_array_schema_uris");
+
+  // Check if array_ls_uris is empty or not, if it is empty, do ls
+  std::vector<URI> uris;
+  if (array_ls_uris->size() == 0) {
+    RETURN_NOT_OK(vfs_->ls(array_uri.add_trailing_slash(), &uris));
```

Comment:
I don't think we need to do a listing on the top level array directory. If the array has lots of fragments this will have a large performance impact. Instead you can:
1) Check for the old style `__array_schema.tdb` file directory with `vfs_->is_file`
2) Do a listing on the `array_uri` + `constants::array_schema_folder_name` + `/`, if the listing is non-empty then there is new style schemas. If its empty then there is no new schema.

Reply:
Pushed. I forgot to push it last week. Changed back to not list on the top level array directory.  Actually that is our original version. I also removed the reuse of the listing result of top level array directory. please have a look at it. thanks